File: src/main/java/frc/robot/subsystems/drivetrain/Drivetrain.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems.drivetrain;

import edu.wpi.first.math.geometry.*;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
import edu.wpi.first.math.kinematics.SwerveModulePosition;
import edu.wpi.first.math.kinematics.SwerveModuleState;
import edu.wpi.first.wpilibj.RobotBase;
import edu.wpi.first.wpilibj.shuffleboard.BuiltInWidgets;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import frc.robot.Constants;
import frc.robot.Constants.DriveConstants;
import frc.robot.Robot;
import frc.robot.RobotContainer;

import java.util.ArrayList;
import java.util.List;

public class Drivetrain extends SubsystemBase {

    // Create MAXSwerveModules
    private final SwerveModules swerveModules;
    // The gyro sensor
    private final Gyro gyro = new Gyro();
    /**
     * Creates a new DriveSubsystem.
     */
    public Drivetrain() {

        this.swerveModules = new SwerveModules(
                new MAXSwerveModule(
                        DriveConstants.FRONT_LEFT_DRIVE_ID,
                        DriveConstants.FRONT_LEFT_TURN_ID,
                        DriveConstants.MODULE_ROTATIONS[0]
                ),
                new MAXSwerveModule(
                        DriveConstants.FRONT_RIGHT_DRIVE_ID,
                        DriveConstants.FRONT_RIGHT_TURN_ID,
                        DriveConstants.MODULE_ROTATIONS[1]
                ),
                new MAXSwerveModule(
                        DriveConstants.REAR_LEFT_DRIVE_ID,
                        DriveConstants.REAR_LEFT_TURN_ID,
                        DriveConstants.MODULE_ROTATIONS[2]
                ),
                new MAXSwerveModule(
                        DriveConstants.REAR_RIGHT_DRIVE_ID,
                        DriveConstants.REAR_RIGHT_TURN_ID,
                        DriveConstants.MODULE_ROTATIONS[3]
                )
        );


        RobotContainer.swerveTab.addNumber("Front Left", swerveModules.frontLeft::getSwerveEncoderPosition).withWidget(BuiltInWidgets.kGraph);
        RobotContainer.swerveTab.addNumber("Front Right", swerveModules.frontRight::getSwerveEncoderPosition).withWidget(BuiltInWidgets.kGraph);
        RobotContainer.swerveTab.addNumber("Back Left", swerveModules.rearLeft::getSwerveEncoderPosition).withWidget(BuiltInWidgets.kGraph);
        RobotContainer.swerveTab.addNumber("Back Right", swerveModules.rearRight::getSwerveEncoderPosition).withWidget(BuiltInWidgets.kGraph);

        RobotContainer.swerveTab.addNumber("Gyro", () -> gyro.getAngle().getRadians()).withWidget(BuiltInWidgets.kGraph);
    }

    @Override
    public void periodic() { 

        //TODO odometry :0

        // RobotContainer.poseEstimation.updateOdometry(
        //         getRotation(),
        //         getModulePositions()
        // );

        gyro.update();
        swerveModules.update();

        logModuleStates("Swerve State", swerveModules.getStates().asArray());
        logModuleStates("Swerve Set State", new SwerveModuleState[]{
                swerveModules.frontLeft.getSetState(),
                swerveModules.frontRight.getSetState(),
                swerveModules.rearLeft.getSetState(),
                swerveModules.rearRight.getSetState()
        });
        SmartDashboard.putNumberArray("Swerve Module Distance", new double[]{
                swerveModules.frontLeft.getPosition().distanceMeters / Constants.ModuleConstants.WHEEL_CIRCUMFERENCE_METERS,
                swerveModules.frontRight.getPosition().distanceMeters / Constants.ModuleConstants.WHEEL_CIRCUMFERENCE_METERS,
                swerveModules.rearLeft.getPosition().distanceMeters / Constants.ModuleConstants.WHEEL_CIRCUMFERENCE_METERS,
                swerveModules.rearRight.getPosition().distanceMeters / Constants.ModuleConstants.WHEEL_CIRCUMFERENCE_METERS});

        SmartDashboard.putNumberArray("Swerve Module Distance Revolutions", new double[]{
                swerveModules.frontLeft.getPosition().distanceMeters,
                swerveModules.frontRight.getPosition().distanceMeters,
                swerveModules.rearLeft.getPosition().distanceMeters,
                swerveModules.rearRight.getPosition().distanceMeters});

        SmartDashboard.putNumber("Pitch", gyro.getRotation3d().getY());
    }

    private void logModuleStates(String key, SwerveModuleState[] swerveModuleStates) {
        List<Double> dataList = new ArrayList<>();
        for (SwerveModuleState swerveModuleState : swerveModuleStates) {
            dataList.add(swerveModuleState.angle.getRadians());
            dataList.add(swerveModuleState.speedMetersPerSecond);
        }
        SmartDashboard.putNumberArray(key,
                dataList.stream().mapToDouble(Double::doubleValue).toArray());
    }

    public Rotation2d getRotation() {
        Rotation2d rot = gyro.getAngle();

        if (DriveConstants.GYRO_REVERSED) {
            rot = rot.unaryMinus();
        }

        return rot;
    }

    public Rotation3d getRotation3d() {
        return gyro.getRotation3d();
    }

    public SwerveModulePosition[] getModulePositions() {
        return swerveModules.getPositions().asArray();
    }

    public SwerveModuleState[] getModuleStates() {
        return swerveModules.getStates().asArray();
    }

    /**
     * Method to drive the drivetrain using chassis speeds.
     *
     * @param speeds The chassis speeds.
     */
    public void drive(ChassisSpeeds speeds) {
        double dt = Robot.kDefaultPeriod;
        Pose2d poseVel = new Pose2d(
                speeds.vxMetersPerSecond * dt,
                speeds.vyMetersPerSecond * dt,
                Rotation2d.fromRadians(speeds.omegaRadiansPerSecond * dt));
        Twist2d twistVel = new Pose2d().log(poseVel);
        ChassisSpeeds correctedSpeeds = new ChassisSpeeds(
                twistVel.dx / dt, twistVel.dy / dt, twistVel.dtheta / dt);

        SwerveModuleState[] swerveModuleStates = DriveConstants.DRIVE_KINEMATICS.toSwerveModuleStates(correctedSpeeds);
        setModuleStates(swerveModuleStates);
    }

    /**
     * Sets the wheels into an X formation to prevent movement.
     */
    public void setX() {
        swerveModules.setDesiredStates(
                new SwerveModules.States(
                        // front left
                        new SwerveModuleState(0, Rotation2d.fromDegrees(45)),
                        // front right
                        new SwerveModuleState(0, Rotation2d.fromDegrees(-45)),
                        // rear left
                        new SwerveModuleState(0, Rotation2d.fromDegrees(-45)),
                        // rear right
                        new SwerveModuleState(0, Rotation2d.fromDegrees(45))
                ).asArray()
        );
    }

    /**
     * Sets the swerve ModuleStates.
     *
     * @param desiredStates The desired SwerveModule states.
     */
    public void setModuleStates(SwerveModuleState[] desiredStates) {
        SwerveDriveKinematics.desaturateWheelSpeeds(desiredStates, DriveConstants.MAX_SPEED_METERS_PER_SECOND);
        swerveModules.setDesiredStates(desiredStates);
    }

    /**
     * Resets the drive encoders to currently read a position of 0.
     */
    public void resetEncoders() {
        swerveModules.resetEncoders();
    }

    /**
     * Zeroes the heading of the robot.
     */
    public void zeroHeading() {
        gyro.reset();
    }

    /**
     * Returns the heading of the robot.
     *
     * @return the robot's heading in degrees, from -180 to 180
     */
    public Rotation2d getHeading() {
        return gyro.getAngle();
    }

    public Translation2d getRelativeVelocity() {
        ChassisSpeeds chassisSpeeds = DriveConstants.DRIVE_KINEMATICS.toChassisSpeeds(getModuleStates());

        return new Translation2d(chassisSpeeds.vxMetersPerSecond, chassisSpeeds.vyMetersPerSecond);
    }

    /**
     * Returns the turn rate of the robot.
     *
     * @return The turn rate of the robot, in degrees per second
     */
    public double getTurnRate() {
        return gyro.getRate().getDegrees();
    }

    public Boolean getNavXStatus() {
        if (RobotBase.isSimulation()) 
            return false;

        return gyro.isConnected();
    }

}

File: src/main/java/frc/robot/subsystems/drivetrain/Drivetrain.kt
package frc.robot.subsystems.drivetrain

import MAXSwerveModuleIO
import edu.wpi.first.math.estimator.SwerveDrivePoseEstimator
import edu.wpi.first.math.geometry.Pose2d
import edu.wpi.first.math.geometry.Translation2d
import edu.wpi.first.math.geometry.Rotation2d
import edu.wpi.first.math.kinematics.ChassisSpeeds
import edu.wpi.first.math.kinematics.SwerveDriveKinematics
import edu.wpi.first.math.kinematics.SwerveModuleState
import edu.wpi.first.wpilibj.RobotBase
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
import edu.wpi.first.wpilibj2.command.CommandScheduler
import edu.wpi.first.wpilibj2.command.Subsystem
import frc.robot.CANDevice
import frc.robot.RobotContainer
import frc.robot.utils.PerCorner
import frc.robot.utils.cornerStatesToChassisSpeeds
import frc.robot.utils.toCornerSwerveModuleStates
import SimSwerveModuleIO
import org.littletonrobotics.junction.Logger
import edu.wpi.first.math.MathShared
import edu.wpi.first.math.util.Units
import ModuleIO
import ModuleIO.Inputs

// A singleton object representing the drivetrain.
object Drivetrain : Subsystem {





    private val modules = if (RobotBase.isReal()) {
        MODULE_CAN_DEVICES.zip(MODULE_POSITIONS).map { (can, pose) ->
            MAXSwerveModuleIO(can.first, can.second, pose.rotation)
        }.zip ( PerCorner.generate { Inputs() } )
    } else {
        PerCorner.generate { SimSwerveModuleIO() }.zip( PerCorner.generate { Inputs() } )
    }



    private val positions = MODULE_POSITIONS

    private val gyro: GyroIO = if (RobotBase.isReal()) {
        NavXGyroIO()
    } else {
        SimGyroIO()
    }

    val gyroInputs = GyroInputs()

    // Create swerve drivetrain kinematics using the translation parts of the module positions.
    private val kinematics = SwerveDriveKinematics(*positions.map(Pose2d::getTranslation).toList().toTypedArray())

    private val poseEstimator = SwerveDrivePoseEstimator(
        kinematics, // swerve drive kinematics
        gyroInputs.rotation.toRotation2d(), // initial gyro rotation
        modules.map { it.second.position }.toTypedArray(), // initial module positions
        Pose2d() // initial pose
        // TODO: add odometry standard deviation
    )

    init {
        CommandScheduler.getInstance().registerSubsystem(this)
    }

    override fun periodic() {
        gyro.updateInputs(gyroInputs)
        // Update each of the modules.
        modules.forEach { (module, input) -> module.updateInputs(input) }
        // Update the estimated position.
        // poseEstimator.update(
        //     gyroInputs.rotation.toRotation2d(), modules.map { it.second.position }.toTypedArray()
        // )


        Logger.getInstance().recordOutput("Odometry/pose", estimatedPose)
        Logger.getInstance().recordOutput("Drive/ModulesState/Measured", *modules.map{it.second.state}.toTypedArray())
        Logger.getInstance().recordOutput("Drive/ModulesState/Desired", *modules.map{it.second.desiredState}.toTypedArray())



        // Log the swerve module states to SmartDashboard.
        SmartDashboard.putNumberArray(
            "Drive/ModuleStates/Measured",
            modules.map {it.second.state}.flatMap { listOf(it.angle.radians, it.speedMetersPerSecond) }
                .toTypedArray()
        )
        SmartDashboard.putNumberArray(
            "Drive/ModuleStates/Desired",
            modules.map {it.second.desiredState}.flatMap { listOf(it.angle.radians, it.speedMetersPerSecond) }
                .toTypedArray()
        )

        // Update the robot position on the SmartDashboard field.
        RobotContainer.field.robotPose = poseEstimator.estimatedPosition
    }

    // The current speed of chassis relative to the ground, assuming that the wheels have traction with the ground.
    val chassisSpeeds: ChassisSpeeds
        get() = kinematics.cornerStatesToChassisSpeeds(modules.map {it.second.state})

    // Set the drivetrain to an X-formation to passively prevent movement in any direction.
    fun brake() {
        modules.zip(MODULE_POSITIONS).forEach { (module, position) ->
            // set the modules to radiate outwards from the chassis origin
            module.second.desiredState = SwerveModuleState(0.0, position.rotation)
        }
    }

    // Set the drivetrain to move with the given chassis speeds.
    //
    // Note that the speeds are relative to the chassis, not the field.
    fun drive(speeds: ChassisSpeeds) {
        
        // TODO: desaturate wheel speeds
        // SwerveDriveKinematics.desaturateWheelSpeeds(speeds, MAX_SPEED)

        kinematics.toCornerSwerveModuleStates(speeds).zip(modules).forEach { (state, module) ->
            module.first.setDesiredState(state) 
        }
    }

    fun zeroGyro() {
        gyro.reset()
    }

    // Get the estimated pose of the drivetrain using the pose estimator.
    val estimatedPose: Pose2d
        get() = poseEstimator.estimatedPosition


}

// Constants
internal val WHEEL_BASE_HALF: Double = Units.inchesToMeters(13.0)

internal val TRACK_WIDTH_HALF: Double = Units.inchesToMeters(14.0)

// Constants
internal val MODULE_POSITIONS = PerCorner(
    frontLeft = Pose2d(Translation2d(WHEEL_BASE_HALF, TRACK_WIDTH_HALF), Rotation2d.fromDegrees(-90.0)), 
    frontRight = Pose2d(Translation2d(WHEEL_BASE_HALF, -TRACK_WIDTH_HALF), Rotation2d.fromDegrees(180.0)), 
    backRight = Pose2d(Translation2d(-WHEEL_BASE_HALF, TRACK_WIDTH_HALF), Rotation2d.fromDegrees(0.0)), 
    backLeft = Pose2d(Translation2d(-WHEEL_BASE_HALF, -TRACK_WIDTH_HALF), Rotation2d.fromDegrees(90.0))
)


internal val MODULE_CAN_DEVICES = PerCorner(
    frontLeft = Pair(CANDevice.FrontLeftDrivingMotor, CANDevice.FrontLeftTurningMotor),
    frontRight = Pair(CANDevice.FrontRightDrivingMotor, CANDevice.FrontRightTurningMotor),
    backRight = Pair(CANDevice.BackRightDrivingMotor, CANDevice.BackRightTurningMotor),
    backLeft = Pair(CANDevice.BackLeftDrivingMotor, CANDevice.BackLeftTurningMotor),
)

